zypp/media/MediaNetworkCommonHandler.cc

```diff
@@ -0,0 +1,81 @@
+/*---------------------------------------------------------------------\
+|                          ____ _   __ __ ___                          |
+|                         |__  / \ / / . \ . \                         |
+|                           / / \ V /|  _/  _/                         |
+|                          / /__ | | | | | |                           |
+|                         /_____||_| |_| |_|                           |
+|                                                                      |
+\---------------------------------------------------------------------*/
+/** \file zypp/media/MediaNetworkCommonHandler.cc
+ *
+*/
+
+#include "MediaNetworkCommonHandler.h"
+
+#include <zypp/ZConfig.h>
+#include <zypp-core/fs/PathInfo.h>
+#include <zypp/base/Logger.h>
+
+#include <fstream>
+
+namespace zypp::media
+{
+  zypp::Url MediaNetworkCommonHandler::findGeoIPRedirect ( const zypp::Url &url )
+  {
+    const auto &hostname = url.getHost();
+    auto geoipFile = ZConfig::instance().geoipCachePath() / hostname ;
+    if ( PathInfo( geoipFile ).isFile() ) {
+
+      MIL << "Found GeoIP file for host: " << hostname << std::endl;
+
+      std::ifstream in( geoipFile.asString() );
+      if (!in.is_open()) {
+        MIL << "Failed to open GeoIP for host: " << hostname << std::endl;
+        return Url();
+      }
+
+      try {
+        std::string newHost;
+        in >> newHost;
+
+        Url newUrl = url;
+        newUrl.setHost( newHost );
+
+        MIL << "Found GeoIP rewrite: " << hostname << " -> " << newHost << " for GeoIP rewrite" << std::endl;
+
+        return newUrl;
+
+      } catch ( const zypp::Exception &e ) {
+        ZYPP_CAUGHT(e);
+        MIL << "Invalid or empty url for GeoIP rewrite" << std::endl;
+      }
+    }
+
+    // no rewrite
+    return Url();
+  }
+
+  Url MediaNetworkCommonHandler::getFileUrl( const Pathname & filename_r ) const
+  {
+    constexpr std::string_view invalidRewrites[] = { "repomd.xml.asc", "repomd.xml.key", "geoip", std::string_view() };
```

Comment:
We should add repomd.xml here as well.
Regional MirrorCache instance does not try to guarantee that its repomd.xml is the latest (it is tricky to guarantee it anyway), so it will redirect requests to repomd.xml tho the main instance.
So it would be better if zypper doesn't use geoip for it as well.
Also media.1/media is tricky file - I am not sure if zypper will still try to access it, but better do not use geoip for it as well (because it is tricky to guarantee freshness of non-versioned files).

Reply:
Done